lib/nerve/reporter/zookeeper.rb

```diff
@@ -145,33 +159,114 @@ def encode_child_name(service)
     end
 
     def zk_delete
-      if @full_key
+      node_path = @full_key_mu.synchronize { @full_key }
+
+      if node_path
         statsd.time('nerve.reporter.zk.delete.elapsed_time', tags: ["zk_cluster:#{@zk_cluster}"]) do
-          @zk.delete(@full_key, :ignore => :no_node)
+          @zk.delete(node_path, :ignore => :no_node)
         end
-        @full_key = nil
+
+        @full_key_mu.synchronize {
+          @full_key = nil
+        }
       end
     end
 
     def zk_create
       # only mkdir_p if the path does not exist
       statsd.time('nerve.reporter.zk.create.elapsed_time', tags: ["zk_cluster:#{@zk_cluster}", "zk_path:#{@zk_path}"]) do
         @zk.mkdir_p(@zk_path) unless @zk.exists?(@zk_path)
-        @full_key = @zk.create(@key_prefix, :data => @data, :mode => @mode)
+
+        node_path = begin
+                      @zk.create(@key_prefix, :data => @data, :mode => @mode)
+                    rescue ::Zookeeper::Exceptions::NodeExists, ZK::Exceptions::NodeExists
+                      # This exception will only occur when not using sequential
+                      # nodes (because sequential nodes are always unique), in which
+                      # case the name is the same as @key_prefix as Zookeeper
+                      # will not append any suffix.
+                      @zk.set(@key_prefix, @data)
+                      log.info "nerve: tried to write node but exists, setting data instead"
+
+                      @key_prefix
+                    end
+
+        @full_key_mu.synchronize {
+          @full_key = node_path
+        }
+        log.info "nerve: wrote new ZK node of type #{@mode} at #{node_path}"
       end
     end
 
-    def zk_save
-      return zk_create unless @full_key
+    def zk_save(node_path)
+      return zk_create unless node_path
 
       begin
         statsd.time('nerve.reporter.zk.save.elapsed_time', tags: ["zk_cluster:#{@zk_cluster}"]) do
-          @zk.set(@full_key, @data)
+          @zk.set(node_path, @data)
+          log.info "nerve: set data on #{node_path}"
         end
       rescue ZK::Exceptions::NoNode
         zk_create
       end
     end
+
+    def start_ttl_renew_thread
+      @ttl_should_exit = false
+      @ttl_thread = nil
+
+      unless @node_ttl.nil? || @mode.to_s.start_with?('ephemeral')
+        @ttl_thread = Thread.new {
+          log.info "nerve: ttl renew: background thread starting"
+          last_run = Time.now - rand(@node_ttl)
+
+          until @ttl_should_exit
```

Comment:
need lock for memory visibility? `@ ttl_should_exit ` is read in a child thread, but set in main thread.